src/logger.js

```diff
@@ -45,4 +69,107 @@ export default once((config) => {
     });
 
     return logger;
-});
+}
+
+// filter out named properties from req object
+function omit(req, blacklist) {
+    return omitBy(req, (value, key) => blacklist.includes(key));
+}
+
+// exclude any health or other ignorable urls
+function skip(config) {
+    return function ignoreUrl(req) {
+        const url = req.originalUrl || req.url;
+        return config.ignoreRouteUrls.includes(url);
+    };
+}
+
+// where morgan connects to winston
+function addStream(logger, config) {
+    return {
+        write: (message) => {
+            const logEntry = JSON.parse(message);
+            return logger.log(config.level, logEntry.message, logEntry);
+        },
+    };
+}
+
+
+@binding('logger')
+@defaults(loggingDefaults)
+class Logger {
+    constructor(graph) {
+        this.config = graph.config.logger;
+        const baseLogger = createLogger(this.config);
+        this.baseLogger = baseLogger;
+        this.requestRules = this.config.requestRules;
+
+        // define custom tokens
+        morgan.token('operation-hash', req => get(req, 'body.extensions.persistentQuery.sha256Hash'));
+        morgan.token('operation-name', req => get(req, 'body.operationName'));
+        morgan.token('user-id', req => get(req, 'user.id'));
+        morgan.token('message', req => req.name || '-');
+        morgan.token('request-id', req => req.id);
+        morgan.token('request-headers', (req) => {
+            const headers = this.config.includeReqHeaders === true
+                ? omit(req.headers, this.config.omitReqProperties) : {};
+            return JSON.stringify(headers);
+        });
+    }
+
+    makeMiddleware() {
```

Comment:
I removed the dependency on the app from this component. This function is something you'd only use in a node context, but you're free to explicitly opt in to this request logging. Maybe we could put our `open-api` specific logic in a similar place: https://github.com/globality-corp/leif/pull/6/files#diff-3c5e700f8a6649b789b7421e57b2f74fR79 ?

@jessemyers what do you think about exposing these as part of the logger vs as functions which accept a graph?

It changes the usage in the node project from:
```
app.use(graph.logger.makeMiddleware())
```

To:
```
app.use(makeMiddleware(graph.logger))
```

Do you think one makes for better boundaries over the other? I guess if we keep adding special "make X" functions to the class the logger will get bloated so maybe the latter option is better.

Reply:
I don't really "get" what this middleware does with the input logger.

Reply:
It generates the per-request audit log:

```
  "length": "1674",
  "message": "graphql",
  "method": "POST",
  "operation-hash": "-",
  "operation-name": "-",
  "request-id": "7363087cdafa4230b150e5545d776af5",
  "request-headers": "{}",
  "response-time": "522.911 ms",
  "status_code": "200",
  "url": "/gql/graphql?",
  "user-id": "-",
  "level": "info"
}
```